File: Easy/0112_path-sum/path-sum.py
#!/bin/python3

from typing import List, Optional
from dataclasses import dataclass
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        if not root.left and not root.right:
            return root.val == targetSum
        return (self.hasPathSum(root.left, targetSum - root.val) or
                self.hasPathSum(root.right, targetSum - root.val))

def buildTree(data: str) -> TreeNode:
    if not data:
        return None

    values = data.split(",")
    if not values or values[0] == "null":
        return None
    
    root = TreeNode(int(values[0]))
    queue = deque([root])
    i = 1

    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] != "null":
            node.left = TreeNode(int(values[i]))
            queue.append(node.left)
        i += 1

        if i < len(values) and values[i] != "null":
            node.right = TreeNode(int(values[i]))
            queue.append(node.right)
        i += 1
    return root

@dataclass
class TestCase:
    tree : str
    sum : int

def main():
    tests: List[TestCase] = [
        TestCase("5,4,8,11,null,13,4,7,2,null,null,null,1", 22),
        TestCase("1,2,3", 5),
        TestCase("", 0),
    ]

    for tc in tests:
        root = buildTree(tc.tree)
        print(f"---\nTree: [{tc.tree}] Sum: {tc.sum}\nhasPathSum: {Solution().hasPathSum(root, tc.sum)}")

if __name__ == "__main__":
    main()

File: Easy/0112_path-sum/path-sum.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)


type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}
	if root.Left == nil && root.Right == nil {
		return root.Val == targetSum
	}
	targetSum -= root.Val
	return hasPathSum(root.Left, targetSum) || hasPathSum(root.Right, targetSum)
}

func buildTree(data string) *TreeNode {
	if data == "" {
		return nil
	}

	values := strings.Split(data, ",")
	if len(values) == 0 || values[0] == "null" {
		return nil
	}

	val, err := strconv.Atoi(values[0])
	if err != nil {
		return nil
	}
	root := &TreeNode{Val: val}

	queue := []*TreeNode{root}
	i := 1

	for len(queue) > 0 && i < len(values) {
		node := queue[0]
		queue = queue[1:]

		if i < len(values) && values[i] != "null" {
			if v, err := strconv.Atoi(values[i]); err == nil {
				node.Left = &TreeNode{Val: v}
				queue = append(queue, node.Left)
			}
		}
		i++

		if i < len(values) && values[i] != "null" {
			if v, err := strconv.Atoi(values[i]); err == nil {
				node.Right = &TreeNode{Val: v}
				queue = append(queue, node.Right)
			}
		}
		i++
	}
	return root
}

type testCase struct {
	tree string
	sum int
}

func main() {
	tests := []testCase{
		{"5,4,8,11,null,13,4,7,2,null,null,null,1", 22},
		{"1,2,3", 5},
		{"", 0},
	}

	for _, tc := range tests {
		root := buildTree(tc.tree)
		fmt.Printf("---\nTree: [%s] Sum: %d\n", tc.tree, tc.sum)
		fmt.Printf("hasPathSum: %t\n", hasPathSum(root, tc.sum))
	}
}
